File: src/main/java/sample/design_pattern/Facade.java
/*
 * Copyright 2013 SatoShun
 *
 * Facade Pattern
 */

package sample.design_pattern;

import java.util.*;


class Basket {
    private List<String> productList;

    public Basket() {
        productList = new ArrayList<String>();
    }

    public void add(String product) {
        productList.add(product);
    }

    public void clear() {
        productList.clear();
    }

    public List<String> getAll() {
        return productList;
    }
}

class Customer {
    private Basket basket;

    public Customer() {
        basket = new Basket();
    }

    public void add(String product) {
        basket.add(product);
    }

    public Map<String,Integer> getAll() {
        List<String> basketContents = basket.getAll();
        Map<String, Integer> map = new HashMap<String, Integer>();

        for (String content : basketContents) {
            if (map.containsKey(content)) {
                map.put(content, map.get(content) + 1);
            } else {
                map.put(content, 1);
            }
        }

        return map;
    }
}


class Buying {
    /*
     * facade method
     */
    public void process() {
        Customer customer = new Customer();

        customer.add("meat");
        customer.add("wine");
        customer.add("meat");
        customer.add("meat");

        System.out.println(customer.getAll());
    }
}



class Facade {
    public static void main(String[] args) {
        Buying buying = new Buying();
        buying.process();
    }
}

File: src/main/scala/sample/design_pattern/Facade.scala
/*
 * Copyright 2013 SatoShun
 *
 * Facade Pattern
 */

package sample.design_pattern;


import scala.collection.mutable.Map

class Basket {
    var productList: List[String] = List()

    def add(product: String) = productList = productList :+ product

    def clear() = productList = List.empty[String]

    def getAll(): List[String] = productList
}


class Customer {
    var basket: Basket = new Basket()

    def add(product: String) = basket.add(product)

    def getAll(): Map[String, Integer] = {
        val basketContents = basket.getAll()
        val result = Map.empty[String, Integer]

        basketContents.foreach({key =>
            key match {
                case key if result.contains(key) => result(key) += 1
                case _ => result(key) = 1
            }
        })
        return result;
    }
}


class Buying {
    def process() = {
        val customer = new Customer()

        customer.add("meat");
        customer.add("wine");
        customer.add("meat");
        customer.add("meat");

        println(customer.getAll())
    }
}

object Facade {
    def main(args: Array[String]) = {
        val buying = new Buying()
        buying.process()
    }
}
